exec/supervisor/plugin.go

```diff
@@ -0,0 +1,222 @@
+	}
+	return names
+}
+
+// GetProcessByName returns an instance of given process
+func (p *Plugin) GetProcessByName(reqName string) pm.ProcessInstance {
+	p.mx.Lock()
+	defer p.mx.Unlock()
+
+	for name, instance := range p.instances {
+		if name == reqName {
+			return instance
+		}
+	}
+	return nil
+}
+
+func (p *Plugin) start(processes []Process) error {
+	pmLogManager := newPmLoggerManager()
+
+	var inError []string
+	for _, process := range processes {
+		if err := p.validate(&process); err != nil {
+			p.Log.Errorf("unable to start process %s: %v", process.Name, err)
+			inError = append(inError, process.Name)
+			continue
+		}
+		processStat := make(chan status.ProcessStatus)
+		pLogger, err := pmLogManager.newPmLogger(process.Name, process.LogFilePath)
+		if err != nil {
+			p.Log.Errorf("error preparing process logger %s: %v", process.Name, err)
+			inError = append(inError, process.Name)
+			continue
+		}
+		instance := p.pm.NewProcess(process.Name, process.BinaryPath,
+			pm.Args(process.Args...),
+			pm.Writer(pLogger, pLogger),
+			pm.Notify(processStat),
+			pm.AutoTerminate())
+		if err := instance.Start(); err != nil {
+			p.Log.Errorf("error starting process %s: %v", process.Name, err)
+			inError = append(inError, process.Name)
+			continue
+		}
+		p.instances[process.Name] = instance
+		p.Log.Debugf("process %s started", process.Name)
+	}
+
+	if len(inError) != 0 {
+		return errors.Errorf("following processes end up with error: %v", inError)
+	}
+	return nil
+}
+
+func (p *Plugin) validate(process *Process) error {
+	if process.Name == "" {
+		// so it will not be logged as empty space
+		process.Name = unnamedProcess
+		return errors.Errorf("name not defined")
+	}
+	if _, ok := p.instances[process.Name]; ok {
+		return errors.Errorf("name %s already exists", process.Name)
+	}
+	if process.BinaryPath == "" {
+		return errors.Errorf("binary %s path not defined", process.Name)
+	}
+	return nil
+}
+
+func (p *Plugin) watch() {
+	for name, instance := range p.instances {
+		// at first check whether the process is still alive
+		if !instance.IsAlive() {
+			// handle the process as if it was terminated
+			p.Log.Warnf("%s does not run, handle as terminated", name)
+			p.handleTerminated(name, instance)
+		} else {
+			go p.watchProcess(name, instance)
+		}
+		p.Log.Debug("process %s watcher started", name)
+	}
+}
+
+func (p *Plugin) watchProcess(name string, instance pm.ProcessInstance) {
+	for {
+		stat := <-instance.GetNotificationChan()
+		if stat == status.Terminated {
+			p.Log.Infof("%s terminated", name)
+			p.handleTerminated(name, instance)
+			return
+		}
+	}
+}
+
+func (p *Plugin) handleTerminated(name string, instance pm.ProcessInstance) {
+	p.mx.Lock()
+	defer p.mx.Unlock()
+
+	// find original config
+	var process Process
+	for _, processCfg := range p.config.Processes {
+		if name == processCfg.Name {
+			process = processCfg
+			break
+		}
+	}
+
+	// execute hooks
+	for _, hook := range process.Hooks {
```

Comment:
We don't want to execute hooks only when terminating program. The hooks should be executed mainly in `watchProcess` for now to handle any change of process state.